### PR TITLE
ParameterCapturing: Remove quotes from string evaluations, include failure reason information make the value nullable

### PR DESCRIPTION
**Pull request by clguiman**
###### Summary

* Updated the captured parameter result so the value parameter and activity id are nullable strings.
* Introduced `ObjectFormatterResult` which includes the evaluated and formatted parameter value and evaluation metadata flags (`IsNull`, `FailedEval`, `UnsupportedEval`, `EvalHasSideEffects`). `ObjectFormatter now returns objects of the new type.
* `EventSourceEmittingProbes.EnterProbe` doesn't request adding quotes to the formatted value anymore. The quotes were added to easily parse the ILogger output, but we don't have that problem anymore.
* Updated `ParameterCapturingEventSource` to send the new evaluation flags as part of the captured paramater.
* Updated `WebApi.Models.CapturedParameter` to include a new optional property: `EvalFailReason`.
* Updated the formatter tests to support the new behavior.

### CODE REVIEW

File: src/Tests/Microsoft.Diagnostics.Monitoring.HostingStartup.UnitTests/ParameterCapturing/ObjectFormatting/Formatters/DebuggerDisplay/DebuggerDisplayFormatterTests.cs [hunk 1]
```diff
@@ -75,10 +84,31 @@ public void GetDebuggerDisplayFormatter_ReturnsWorkingFormatter()
...
-            string formattedResult = factoryResult.Formatter(testObj);
+            ObjectFormatterResult formattedResult = factoryResult.Formatter(testObj);
+
+            // Assert
+            Assert.Equal(FormattableString.Invariant($"'Count = {testObj.Count}'"), formattedResult.FormattedValue);
+            Assert.Equal(ParameterEvaluationFlags.None, formattedResult.Flags);
+        }
+
+        [Fact]
+        public void GetDebuggerDisplayFormatter_FormatsNull()
```
Comment (schmittjoseph) [inline src/Tests/Microsoft.Diagnostics.Monitoring.HostingStartup.UnitTests/ParameterCapturing/ObjectFormatting/Formatters/DebuggerDisplay/DebuggerDisplayFormatterTests.cs:95]: This test is really testing behavior of `ExpressionBinder` and not `GetDebuggerDisplayFormatter`. Let's update the `ExpressionBinderTests` instead with this new null case.

I'd recommend just adding a new `InlineData` for this test case (and updating `DebuggerDisplayClass` with a new method/property that is null.
https://github.com/dotnet/dotnet-monitor/blob/efbb575789e1137fb4416a33a3279ab64a26d38a/src/Tests/Microsoft.Diagnostics.Monitoring.HostingStartup.UnitTests/ParameterCapturing/ObjectFormatting/Formatters/DebuggerDisplay/ExpressionBinderTests.cs#L55

Comment (schmittjoseph) [reply src/Tests/Microsoft.Diagnostics.Monitoring.HostingStartup.UnitTests/ParameterCapturing/ObjectFormatting/Formatters/DebuggerDisplay/DebuggerDisplayFormatterTests.cs:95]: Nevermind, looks like I misread what got changed here!

File: src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ObjectFormatting/Formatters/DebuggerDisplay/ExpressionBinder.cs [hunk 2]
```diff
@@ -26,7 +27,7 @@ internal record struct ExpressionEvaluator(BoundEvaluator Evaluate, Type ReturnT
 
             if (debuggerDisplay.Expressions.Count == 0)
             {
-                return (_, _) => debuggerDisplay.FormatString;
+                return (_, _) => new() { FormattedValue = debuggerDisplay.FormatString, Flags = ParameterEvaluationFlags.None };
```
[the remark this answers is not in the document]
Comment (jander-msft) [reply src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ObjectFormatting/Formatters/DebuggerDisplay/ExpressionBinder.cs:30]: There's only one caller, so it should be able to take the formatted string and turn it into a `ObjectFormatterFunc`.

[the remark this answers is not in the document]
Comment (clguiman) [reply src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ObjectFormatting/Formatters/DebuggerDisplay/ExpressionBinder.cs:30]: I can't return a `string?` because the `ObjectFormatterFunc` takes in a `formatSpecifier` and the `object` as input and that's used to construct the final string (see lines 52 and 62).
I could return a new delegate type that takes an `object` and `FormatSpecifier` as input and returns a `string` instead of `ObjectFormatterResult`, but that seems like a very specific type that's only used in a single place.
What do you think?

File: documentation/api/definitions.md [hunk 3]
```diff
@@ -116,6 +116,8 @@ Object describing a captured parameter.
 | `value` | string | The parameter value. |
 | `typeName` | string | The parameter type name. |
 | `moduleName` | string | The parameter type module name. |
+| `evalFailReason` | string | The reason why evaluation failed. If missing the evaluation was successful. |
+| `isNull` | bool | Whether the evaluated parameter is null (value is omitted if false). |
```
Comment (schmittjoseph) [inline documentation/api/definitions.md:120]: It seems like it might make more sense to remove this field and simply make the `value` field a nullable-string instead. If the value is null, then we use null literal to represent it.

Comment (clguiman) [reply documentation/api/definitions.md]: That makes sense. I've updated the code to make the parameter value and activityId nullable strings. Nullable types are not enabled on the project level, so I had to enable them in the files I touched and I also had to include a workaround for the `required` modifier in .NET 6

File: src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ObjectFormatting/ObjectFormatter.cs [hunk 4]
```diff
@@ -16,6 +17,19 @@ internal enum FormatSpecifier
         NoSideEffects = 2
     }
 
+    internal readonly struct ObjectFormatterResult
+    {
+        public string FormattedValue { get; init; }
+
+        public ParameterEvaluationFlags Flags { get; init; }
+
+        public static readonly ObjectFormatterResult Empty = new() { FormattedValue = string.Empty, Flags = ParameterEvaluationFlags.None };
```
Comment (schmittjoseph) [inline src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ObjectFormatting/ObjectFormatter.cs:26]: Unused?

File: src/Tools/dotnet-monitor/ParameterCapturing/ParameterCapturingEvents.cs [hunk 5]
```diff
@@ -88,8 +90,19 @@ public static class CapturedParameterPayloads
...
             public const int ParameterValue = 5;
-            public const int ParameterAttributes = 6;
-            public const int ParameterTypeIsByRef = 7;
+            public const int ParameterValueEvaluationFlags = 6;
+            public const int ParameterAttributes = 7;
+            public const int ParameterTypeIsByRef = 8;
+        }
+
+        [Flags]
+        public enum ParameterEvaluationFlags
```
Comment (schmittjoseph) [inline src/Tools/dotnet-monitor/ParameterCapturing/ParameterCapturingEvents.cs:99]: It looks like we don't use more than 1 flag at a given time (i.e. they are exclusive), let's make this a regular enum instead?

File: src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ObjectFormatting/ObjectFormatter.cs [hunk 6]
```diff
@@ -16,6 +17,19 @@ internal enum FormatSpecifier
         NoSideEffects = 2
     }
 
+    internal readonly struct ObjectFormatterResult
+    {
+        public string FormattedValue { get; init; }
+
+        public ParameterEvaluationFlags Flags { get; init; }
```
Comment (schmittjoseph) [inline src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ObjectFormatting/ObjectFormatter.cs:24]: Let's have a default value of `ParameterEvaluationFlags.None` instead of requiring everyone to set it? We have to explicitly set this flag to `None` in several places when IMO it should be the default value

File: src/Tools/dotnet-monitor/ParameterCapturing/CapturedParameters.cs [hunk 7]
```diff
@@ -28,7 +30,7 @@ public void AddParameter(ParameterInfo parameter)
             _parameters.Add(parameter);
         }
 
-        public string ActivityId { get; }
+        public string? ActivityId { get; }
 
         public ActivityIdFormat ActivityIdFormat { get; }
```
Comment (schmittjoseph) [inline src/Tools/dotnet-monitor/ParameterCapturing/CapturedParameters.cs:35]: Shouldn't this also be nullable now?

Comment (clguiman) [reply src/Tools/dotnet-monitor/ParameterCapturing/CapturedParameters.cs:35]: The `ActivityIdFormat` enum has an `Unknown` value. From practice I noticed that's set when `ActivityId` is `null`

File: src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ObjectFormatting/Formatters/DebuggerDisplay/ExpressionBinder.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (clguiman) [reply src/Microsoft.Diagnostics.Monitoring.HostingStartup/ParameterCapturing/ObjectFormatting/Formatters/DebuggerDisplay/ExpressionBinder.cs]: I updated the code on Joe's suggestion to default the eval result to `Success`, so there's no need to always set it.

**Review by schmittjoseph — APPROVED** [review]
